src/connectors/impls/gcs/writer.rs

```diff
@@ -0,0 +1,682 @@
+    #[serde(default = "default_connect_timeout")]
+    #[cfg_attr(test, allow(unused))]
+    connect_timeout: u64,
+    #[serde(default = "default_buffer_size")]
+    buffer_size: usize,
+}
+
+fn default_endpoint() -> String {
+    "https://storage.googleapis.com/upload/storage/v1".to_string()
+}
+
+fn default_connect_timeout() -> u64 {
+    10_000_000_000
+}
+
+fn default_buffer_size() -> usize {
+    1024 * 1024 * 8 // 8MB - the recommended minimum
+}
+
+impl ConfigImpl for Config {}
+
+#[derive(Debug, Default)]
+pub(crate) struct Builder {}
+
+#[async_trait::async_trait]
+impl ConnectorBuilder for Builder {
+    fn connector_type(&self) -> ConnectorType {
+        ConnectorType("gcs_writer".into())
+    }
+
+    async fn build_cfg(
+        &self,
+        _alias: &str,
+        _config: &ConnectorConfig,
+        connector_config: &Value,
+        _kill_switch: &KillSwitch,
+    ) -> Result<Box<dyn Connector>> {
+        let config = Config::new(connector_config)?;
+
+        if config.buffer_size % (256 * 1024) != 0 {
+            return Err("Buffer size must be a multiple of 256kiB".into());
+        }
+        let url = Url::<HttpsDefaults>::parse(&config.endpoint)?;
+
+        Ok(Box::new(GCSWriterConnector { config, url }))
+    }
+}
+
+struct GCSWriterConnector {
+    config: Config,
+    url: Url<HttpsDefaults>,
+}
+
+#[async_trait::async_trait]
+impl Connector for GCSWriterConnector {
+    async fn create_sink(
+        &mut self,
+        sink_context: SinkContext,
+        builder: SinkManagerBuilder,
+    ) -> Result<Option<SinkAddr>> {
+        let sink = GCSWriterSink {
+            client: None,
+            url: self.url.clone(),
+            config: self.config.clone(),
+            buffers: ChunkedBuffer::new(self.config.buffer_size),
+            current_name: None,
+            current_session_url: None,
+        };
+
+        let addr = builder.spawn(sink, sink_context)?;
+        Ok(Some(addr))
+    }
+
+    fn codec_requirements(&self) -> CodecReq {
+        CodecReq::Required
+    }
+}
+
+struct ChunkedBuffer {
+    data: Vec<u8>,
+    block_size: usize,
+    buffer_start: usize,
+}
+
+impl ChunkedBuffer {
+    pub fn new(size: usize) -> Self {
+        Self {
+            data: Vec::with_capacity(size * 2),
+            block_size: size,
+            buffer_start: 0,
+        }
+    }
+
+    pub fn mark_done_until(&mut self, position: usize) -> Result<()> {
+        if position < self.buffer_start {
+            return Err("Buffer was marked as done at index which is not in memory anymore".into());
+        }
+
+        let bytes_to_remove = position - self.buffer_start;
+        self.data = Vec::from(self.data.get(bytes_to_remove..).ok_or(
+            ErrorKind::GoogleCloudStorageError("Not enough data in the buffer"),
+        )?);
+        self.buffer_start += bytes_to_remove;
+
+        Ok(())
+    }
+
+    pub fn read_current_block(&self) -> Option<&[u8]> {
+        self.data.get(..self.block_size)
```

Comment:
they way the data is used below (the `Body` impl clones the slice again), it might be more efficient to use https://doc.rust-lang.org/std/vec/struct.Vec.html#method.split_off and return self and assign the returned vec to `self.data` (don't forget to set the capacity to `self.block_size * 2` afterwards).

To avoid `split_off` panicking, you could use it like this `self.data.split_off(self.block_size.min(self.data.len())`

Reply:
I'm afraid we can't do this, because there's no guarantee that google will persist all of the data that we're sending to it. I.e. we might send 256kiB, but google can persists only 128kiB of that (or 0), so we need to keep it in memory until we receive a response and know how much we can throw away. 

```
If the request succeeds, the server responds with 308 Resume Incomplete. The response contains a [Range header](https://cloud.google.com/storage/docs/json_api/v1/parameters#range). Use the upper value in this header to determine where to start the next chunk. You should not assume that the server received all bytes sent in the request.
```
(https://cloud.google.com/storage/docs/performing-resumable-uploads#chunked-upload)

Reply:
Got it, that's why we need to have the actual cutting and removal of old stuff separate. Good point! :)